File: decodeLogService/rabbitmq/producer.js

var amqp = require('amqplib');
const connectDB = require('../Service/dbConfig')
const web3 = require('../Service/web3Config')

async function produceLog(startBlockNumber, endBlockNumber, options = 0) {
  try {
    startBlockNumber = startBlockNumber ? startBlockNumber : 0;
    endBlockNumber = endBlockNumber ? endBlockNumber : await web3.eth.getBlockNumber();

    const client = await connectDB();
    const connection = await amqp.connect('amqp://localhost')
    const channel = await connection.createChannel();
    channel.assertQueue("logs_queue_lazy_db", {
      durable: true,
      queueMode: 'lazy'
    })
    let log_data;
    if(options==0) {
      log_data = await client.query(`SELECT log_index, transaction_hash, address, data, topic0, topic1, topic2, topic3,block_number FROM logs\
                                        where block_number >= ${startBlockNumber} AND block_number <= ${endBlockNumber} AND is_decoded = false AND decode_failed = false`);
    } else if (options == 1) {
      log_data = await client.query(`SELECT log_index, transaction_hash, address, data, topic0, topic1, topic2, topic3,block_number FROM logs\
                                        where block_number >= ${startBlockNumber} AND block_number <= ${endBlockNumber} AND decode_failed = true`);
    } else if (options == 2) {
      log_data = await client.query(`SELECT log_index, transaction_hash, address, data, topic0, topic1, topic2, topic3,block_number FROM logs\
                                        where block_number >= ${startBlockNumber} AND block_number <= ${endBlockNumber}`);
    }
    let groupedObject = log_data.rows.reduce((acc, obj) => {
      const key = obj.transaction_hash;
      if (!acc[key]) {
        acc[key] = []
      }
      acc[key].push(obj)
      return acc;
    }, {})

    let temp = {};
    for (let logs in groupedObject) {
      temp[logs] = groupedObject[logs];
      console.log('buffer: ', Buffer.from(JSON.stringify(temp)))
      // for (let i = 0; i <= 30; i++)
      channel.sendToQueue('logs_queue_lazy_db', Buffer.from(JSON.stringify(temp)), {persistent: true})
      temp = {};
      console.log(" [x] Sent '%s'", logs);
    }
    console.log('Number logs: ', log_data.rows.length)
    } catch (error) {
      console.error(error)
    }
}
produceLog(0, 9000000, 0)
  // produceLog(7001350, 7009000, 1)
// setTimeout(() => {
//   produceLog(0, 7009000)
// }, 10000)

module.exports = produceLog;

File: decodeLogService/rabbitmq/producer.py
import pika

connection = pika.BlockingConnection(pika.ConnectionParameters('localhost'))
channel = connection.channel()

channel.queue_declare(queue='logs_queue', durable=True)

import json

data = {'0x20bb043a5ec9e1cfb977a3ace8c5df8c828308694acaef82f5452b576575a322':
[{'type': 'log', 'log_index': 5, 'transaction_hash': '0x20bb043a5ec9e1cfb977a3ace8c5df8c828308694acaef82f5452b576575a322', 'transaction_index': 2, 'block_hash': '0x8cb5f7c3500b63c324c0255eefb8e3a7bc09ed87835737060e8c665d77b7479a', 'block_number': 8019670, 'address': '0x526d7d217a7996932ae54860c7a9ca79547817c0', 'data': '0xec471d8c985c17e9da641543717ee678d0dd93bed73deb6330424541d410ebef0000000000000000000000001fae0b37ff268e6a80ef8ad1a928fe2d4468efc800000000000000000000000000000000000000000000000000000000000000800000000000000000000000000000000000000000000000000000000000000001000000000000000000000000000000000000000000000000000000000000000b4769c3b22053e1bb916e67000000000000000000000000000000000000000000', 'topics': ['0xedccefa262e9e9aa013e50d7172f89c38ca112ac3f1b34f42f4b07e74f680fe5'], 'decode': 'None'}]}

print(type(data))
channel.basic_publish(exchange='',
                      routing_key='logs_queue',
                      body=json.dumps(data))
print(" [x] Sent 'Hello World!'")


connection.close()